### PR TITLE
Test vectors with failure cases

### PR DESCRIPTION
**Pull request by wtoorop**
This includes Pieter Lexis' test vectors and has some additional failure cases added.

### CODE REVIEW

**Comment by wtoorop** [comment]
More commits with more failure cases (from @TCY16) will follow shortly...

File: draft-ietf-dnsop-svcb-https.md
Comment (bemasc) [inline draft-ietf-dnsop-svcb-https.md:1942]: No, this record is perfectly valid.  "automatically mandatory" means "if these keys are present, clients must act as if they were in the `mandatory=...` list".

File: draft-ietf-dnsop-svcb-https.md
Comment (bemasc) [inline draft-ietf-dnsop-svcb-https.md:1950]: As your text notes, this is a "should not", so the record is not invalid.

File: draft-ietf-dnsop-svcb-https.md
Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md]: Removed

File: draft-ietf-dnsop-svcb-https.md
Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md]: and removed, thanks for the feedback Ben!

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1907]:
```suggestion
This example has multiple instances of the same
```

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1920]:
```suggestion
The "no-default-alpn" SvcParamKey value MUST be empty ({{alpn-key}}).
```

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1928]:
```suggestion
The "mandatory" SvcParamKey MUST not be included in mandatory
```

File: draft-ietf-dnsop-svcb-https.md
Comment (marka63) [inline draft-ietf-dnsop-svcb-https.md:1917]: These should be done in individual records rather than 1 record to ensure proper error detection.

Comment (marka63) [inline draft-ietf-dnsop-svcb-https.md:1917]: Add keyXXX forms as well.

File: draft-ietf-dnsop-svcb-https.md
Comment (marka63) [inline draft-ietf-dnsop-svcb-https.md:1817]: This may need to done in multi-line format to allow it to fit nicely into .txt format.

File: draft-ietf-dnsop-svcb-https.md
Comment (marka63) [inline draft-ietf-dnsop-svcb-https.md:1917]: add mandatory to the list.

**Comment by bemasc** [comment]
Do we have an example of a published RFC with such a detailed catalog of invalid test vectors?  If not, perhaps it's inappropriate for a standards document.  After all, we are not trying to ship a complete test suite.

**Comment by wtoorop** [comment]
Hi @bemasc , I can't think of a RFC from head, but I know that two ZONEMD implementations completely missed that they should not try and *fail* verification with two ZONEMD RR's with the same algorithm. A test-vector for the failure case would probably have prevented that, as implementations tend to copy and use those in their test suits. What would be a good reason to not include them? Do you want me to ask the workgroup again about whether they think it should be included or not?

Also, @pieterlexis' test-vectors already payed off! Because both the ldns implementation and the Net::DNS implementation missed that alpn values have to be escaped and unescaped twice!

File: draft-ietf-dnsop-svcb-https.md
Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md]: resolved

Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md]: resolved

Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md]: resolved

File: draft-ietf-dnsop-svcb-https.md
Comment (TCY16) [reply draft-ietf-dnsop-svcb-https.md:1817]: This _should_ fit the .txt file as it is lined out to be on the exact maximum

**Comment by TCY16** [comment]
Hi @bemasc,

I do get what you’re saying about the complete test suite and agree that some are slightly triviaal, though I think there is still added value of including at least some of the tests. 
I started this list mainly for the NSD test suite, but as both you and @marka64 pointed out, there were some mistakes/improvements. By adding a subset of these tests me and @wtoorop hope that other implementers are less prone to incorrect implementations.

Do you disagree?

**Comment by moonshiner** [comment]
I agree with @TCY16 - we should more inclusive with the examples.  They seem to trivial to us, but we are not the people whom will reading this and attempting to updating their DNS zone data.   Trivial examples are the turtles folks will use to reach the next set of turtles. 

(speaking only for myself)

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1922]:
```suggestion
    example.com.   SVCB   1 foo.example.com. key123
```

File: draft-ietf-dnsop-svcb-https.md
Comment (pieterlexis) [inline draft-ietf-dnsop-svcb-https.md:1916]: Another failure case would be:

```
     example.com.   SVCB   1 foo.example.com. mandatory=
```

File: draft-ietf-dnsop-svcb-https.md
Comment (bemasc) [inline draft-ietf-dnsop-svcb-https.md:1920]: This is a tricky one.  The param indicates a valid SVCB `echconfig` parameter whose value is "" (which is what you get when you base64-decode "").

Of course, "" is currently not a valid ECHConfigList, and is unlikely to be, but the ECH specification is not final.  In any event, I don't think there's an obligation for zone file parsers to check whether the ECHConfigList is valid, which may be difficult if they're not linked to an ECH-capable TLS library.

File: draft-ietf-dnsop-svcb-https.md
Comment (bemasc) [inline draft-ietf-dnsop-svcb-https.md:1922]: This is definitely allowed.  `key123` could someday be registered as a key that uses an empty value.  Until it's registered, zone files can't apply any limits.

**Review by bemasc — COMMENTED** [review]
OK, it sounds like there is a lot of interest in including invalid RDATA.

File: draft-ietf-dnsop-svcb-https.md
Comment (pieterlexis) [reply draft-ietf-dnsop-svcb-https.md:1922]: I know an earlier draft mentioned that empty generic keys were not allowed and only no-default-alpn was defined as having no value?

Comment (bemasc) [reply draft-ietf-dnsop-svcb-https.md:1922]: I don't recall a limit like that.  Perhaps an early draft implied that this should be spelled `key123=""`, but there is no such requirement now.

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1920]:
```suggestion
```
Allow for empty echconfig lists. Because ECH spec is not final.

**Review by wtoorop — COMMENTED** [review]
Suggestions based on review comments from @bemasc

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [reply draft-ietf-dnsop-svcb-https.md]: Removed that one.

File: draft-ietf-dnsop-svcb-https.md
Comment (wtoorop) [reply draft-ietf-dnsop-svcb-https.md]: Removed this line too.

**Review by bemasc — APPROVED** [review]
Waiting for coauthor approval.

**Comment by Habbie** [comment]
> Do we have an example of a published RFC with such a detailed catalog of invalid test vectors? If not, perhaps it's inappropriate for a standards document. After all, we are not trying to ship a complete test suite.

The few positive examples in CSYNC really helped me in implementing the type, and I wouldn't have minded some 'negative' examples in there too.

In general, I feel RFCs generally have too few test vectors (often zero), not too many.

I think shipping a complete test suite would be an excellent way to lead by example!

**Comment by enygren** [comment]
Opened issue #315  to test the test vectors.  It would be great if people could confirm there if they've tested the test vectors.

**Pull request merged**
